span-builder/service.py

```diff
@@ -0,0 +1,203 @@
+                if (
+                    int(other["startTimeUnixNano"]) > parent_starttime
+                    and int(other["endTimeUnixNano"]) < parent_endtime
+                ):
+                    span["parentSpanId"] = other["spanId"]
+                    parent_starttime = int(other["startTimeUnixNano"])
+                    parent_endtime = int(other["endTimeUnixNano"])
+    return payload
+
+
+def _is_trace_complete(val_trace_id):
+    # TODO: determine when a trace is complete
+    # criteria: function START should be closed by a function END
+    #
+    # However, here's a counter example:
+    # Let's say we have trace1, comprising of span1 and span2,
+    # which are non-overlapping in time.
+    # We can have children like span1-1, span1-2, whatever.
+    # When the span builder receives only span1-1, span1-2, and span1,
+    # what makes the span builder wait for span2?
+    # Because it can just send trace1->span1->{span1-1, span1-2} and that would be a valid trace
+    # In this case we'll be missing span2 information
+
+    with data_store_locks[val_trace_id]:
+        for val_node_id in data_store[val_trace_id]:
+            for val_thread_id in data_store[val_trace_id][val_node_id]:
+                for val_span_name in data_store[val_trace_id][val_node_id][
+                    val_thread_id
+                ]:
+                    if (
+                        Stage.START.name
+                        not in data_store[val_trace_id][val_node_id][val_thread_id][
+                            val_span_name
+                        ]
+                        or Stage.STOP.name
+                        not in data_store[val_trace_id][val_node_id][val_thread_id][
+                            val_span_name
+                        ]
+                    ):
+                        return False
+    return True
+
+
+def _construct_trace_object(val_trace_id):
+    payload = copy.deepcopy(STARTER_SPAN)
+    with data_store_locks[val_trace_id]:
+        for val_node_id in data_store[val_trace_id]:
+            for val_thread_id in data_store[val_trace_id][val_node_id]:
+                for val_span_name in data_store[val_trace_id][val_node_id][
+                    val_thread_id
+                ]:
+                    span_start_log = data_store[val_trace_id][val_node_id][
+                        val_thread_id
+                    ][val_span_name][Stage.START.name]
+                    span_stop_log = data_store[val_trace_id][val_node_id][
+                        val_thread_id
+                    ][val_span_name][Stage.STOP.name]
+                    span = _construct_span_object_from_log(
+                        span_start_log, span_stop_log
+                    )
+                    payload["resourceSpans"][0]["scopeSpans"][0]["spans"].append(span)
+    # TODO: We should also delete the raw logs after we deem it complete and construct a trace object from it
+    return payload
```

Comment:
Nice. For our purpose, I think this is fine. We can optimize later. Maybe some sort of tree could be more efficient but I don't think it matters that much atm.